include/pybind11/cast.h

```diff
@@ -1027,6 +1027,8 @@ struct type_caster<T, enable_if_t<std::is_arithmetic<T>::value && !is_std_char_t
                 return false;
         } else if (PyFloat_Check(src.ptr())) {
             return false;
+        } else if (!convert && !PYBIND11_LONG_CHECK(src.ptr()) && !PyIndex_Check(src.ptr())) {
```

Comment:
Maybe reverse the order of `PYBIND11_LONG_CHECK` and `PyIndex_Check`?
`PyIndex_Check` is a very simple lookup of struct members:
https://github.com/python/cpython/blob/2edfc86f69d8a74f4821974678f664ff94a9dc22/Include/internal/pycore_abstract.h#L13
`PyLong_Check` seems more involved:
https://github.com/python/cpython/blob/2edfc86f69d8a74f4821974678f664ff94a9dc22/Include/longobject.h#L14

Reply:
Makes sense, yes. Will do.

Reply:
In the end, `PyLong_Check` is also pretty trivial, though: https://github.com/python/cpython/blob/2edfc86f69d8a74f4821974678f664ff94a9dc22/Include/object.h#L671-L684